### PR TITLE
Undo assertion on STANDALONE_WASM not working with memory growth

### LINKED ISSUE #9587: Build regression: "STANDALONE_WASM does not support memory growth yet"
The following command line used to work a few months ago - I think this was with 1.38.40:

`
emcc src/vertexcodec.cpp src/indexcodec.cpp -O3 -DNDEBUG -s EXPORTED_FUNCTIONS='["_meshopt_decodeVertexBuffer", "_meshopt_decodeIndexBuffer"]' -s ALLOW_MEMORY_GROWTH=1 -s TOTAL_STACK=24576 -s TOTAL_MEMORY=65536 -o build/meshopt_decoder.wasm
`

After uprading to 1.38.47, I now get the following error:

`
shared:ERROR: STANDALONE_WASM does not support memory growth yet
`

Memory growth worked correctly before the update. Any pointers as to how to fix this would be appreciated.

### PR DESCRIPTION
**Pull request by kripken**
While it isn't 100% standalone (there is an import to grow the memory), it is usable for people that implement that import, so don't break them.

fixes #9587

### CODE REVIEW

**Comment by zeux** [comment]
Just a minor comment, "there is an import to grow the memory" - there actually isn't :) emscripten_resize_heap is implemented in standalone_wasm.c without any callouts to JS.

**Comment by kripken** [comment]
Thanks @zeux! Actually that's more than a minor comment, it showed I missed something important here, and had to rework this PR.

We don't support memory growth in pure wasm runtimes yet, because of https://github.com/WebAssembly/WASI/issues/82 My goal was for this PR to make it work outside of those runtimes - that is, when standalone mode is enabled, but still running with the JS. And my assumption was that wasm runtimes would show an error on `emscripten_resize_heap` being missing, would let people know why it doesn't work (because extra non-wasi stuff is needed). But you're right, that function was already included in the wasm...

I rewrote this PR to do things properly, and added a test for growth as well:

 * Remove the assertion on memory growth not being allowed in standalone mode.
 * Add `emscripten_notify_memory_growth` which is just called to notify. That will show an error in wasm runtimes that don't support it.
 * Add a variation on libstandalone that calls that notification if growth is enabled. That lets programs not using growth still be 100% pure.
 * Add a test for memory growth in "impure" standalone wasm (i.e., standalone mode does not emit a 100% pure wasm yet).
 * Fix some bugs in `emscripten_resize_heap` and the preamble memory loading code, that the test uncovered.

**Comment by VirtualTim** [comment]
Hey would this change also fix #9190?

**Comment by kripken** [comment]
@VirtualTim this is separate from that, as this PR is just for standalone wasm mode - in which there is no JS. We don't have support for pthreads in that case, but on the plus side, without JS the problem in #9190 won't exist, if/when we do have standalone wasm pthreads.

**Comment by kripken** [comment]
I also added docs for the new API.

@sbc100  maybe you want to take another look here, as it changed since the lgtm?

File: tools/system_libs.py [hunk 1]
```diff
@@ -1185,6 +1185,33 @@ class libstandalonewasm(MuslInternalLibrary):
...
   src_dir = ['system', 'lib']
 
+  def __init__(self, **kwargs):
+    self.is_mem_grow = kwargs.pop('is_mem_grow')
+    super(libstandalonewasm, self).__init__(**kwargs)
+
+  def get_base_name(self):
+    name = super(libstandalonewasm, self).get_base_name()
+    if self.is_mem_grow:
+      name += '-mem_grow'
```
Comment (sbc100) [inline tools/system_libs.py]: Seems odd to mix hyphen and underscore here.

Comment (kripken) [reply tools/system_libs.py]: Good point, I'll make it `memgrow` (no underscore).

**Pull request merged**